### ISSUE TITLE
HyLEAR very slow & not learning properly

### ISSUE THREAD

**Issue by DP1997** [dikshant2210/Carla-CTS02#3]
Hi @dikshant2210
I have been trying to get HyLEAR to run over the past couple weeks, but without success so far.
While doing so I noticed that it trains extremely slow, i.e. it roughly takes 3 days for 500 episodes, which seems awfully slow (other methods manage 5000 episodes in this time).
Furthermore, I am also tracking its training progress live and even after 1500 episodes it reaches the goal only 5% of the time, which doesn't seem right either.
I tried with both the hyperparameter sets specified in the respective config file:
https://github.com/dikshant2210/Carla-CTS02/blob/5e83b7f07a15e41a2129a28037314a4f895cad1d/SAC/sac_discrete/config/sacd.yaml#L1-L17 
and as the provided default values of the hylear_agent.py:
https://github.com/dikshant2210/Carla-CTS02/blob/5e83b7f07a15e41a2129a28037314a4f895cad1d/hylear/hylear_agent.py#L20-L26

Since I am not familiar with SAC agents, I would very much like to rely on your expertise.
Does anything come to mind what could be causing this?
Or do you have any suggestions what I should be trying? 

Any help would be greatly appreciated!

Thanks ~ Donald

**Comment by dikshant2210** [comment]
From the description few things come to mind:
1. One of the reasons behind training being slow is that Carla cannot run in headless mode i.e. even after giving no render flag, Carla still renders everything behind the scenes and just does not display it. Even with this problem, 500 episodes in three days is too slow. My guess is probably Carla is not utilizing the correct GPU to run. Sometimes an extra nvidia flag is required to run Carla. Please ensure Carla is running using the correct GPU.
2. The second possible reason for slow training is the agent is not learning at all (which might be the case here). This is quite surprising as hylear utilizes hybrid Planner policy to train i.e., actions are sampled using hybrid planner policy and not randomly. Please check if policy from hybrid planner is correct and is being used to sample action policy during training. You can also change the initialization of categorical distribution based on hybrid planner policy to expedite training i.e. use (0.9, 0.05, 0.05) rather than current (0.8, 0.1, 0.1). Please remember this comes with tradeoff (typical exploration vs exploitation problem of off-policy learning).
3. The third possible reason for slow training is that the car is not reaching its goal. Basically, the car is stationary or moving extremely slowly which means all episodes run for maximum time steps (500 or 600) which makes the training slower (with a good policy the car should arrive at the goal within 300 time steps). This problem should resolve itself if the second point is rectified.

Note: In general training DRL agents on autonomous navigation are notoriously unstable. It is quite possible for the model to diverge after converging. Therefore training over a large number of episodes is not always guaranteed to produce better results. One thing that I remember doing was to log the performance over a subset of validation data at regular intervals (e.g., after every 50 or 100 episodes) and then choose the checkpoint that has the highest performance on the subset of the validation set.

**Comment by DP1997** [comment]
> From the description few things come to mind:
> 
> 1. One of the reasons behind training being slow is that Carla cannot run in headless mode i.e. even after giving no render flag, Carla still renders everything behind the scenes and just does not display it. Even with this problem, 500 episodes in three days is too slow. My guess is probably Carla is not utilizing the correct GPU to run. Sometimes an extra nvidia flag is required to run Carla. Please ensure Carla is running using the correct GPU.
> 2. The second possible reason for slow training is the agent is not learning at all (which might be the case here). This is quite surprising as hylear utilizes hybrid Planner policy to train i.e., actions are sampled using hybrid planner policy and not randomly. Please check if policy from hybrid planner is correct and is being used to sample action policy during training. You can also change the initialization of categorical distribution based on hybrid planner policy to expedite training i.e. use (0.9, 0.05, 0.05) rather than current (0.8, 0.1, 0.1). Please remember this comes with tradeoff (typical exploration vs exploitation problem of off-policy learning).
> 3. The third possible reason for slow training is that the car is not reaching its goal. Basically, the car is stationary or moving extremely slowly which means all episodes run for maximum time steps (500 or 600) which makes the training slower (with a good policy the car should arrive at the goal within 300 time steps). This problem should resolve itself if the second point is rectified.
> 
> Note: In general training DRL agents on autonomous navigation are notoriously unstable. It is quite possible for the model to diverge after converging. Therefore training over a large number of episodes is not always guaranteed to produce better results. One thing that I remember doing was to log the performance over a subset of validation data at regular intervals (e.g., after every 50 or 100 episodes) and then choose the checkpoint that has the highest performance on the subset of the validation set.

Hi @dikshant2210 ,
thanks a lot for your detailed answer and insightful tips  - it is much appreciated!
I will look into the things you mentioned and hope that this will resolve the issue :)

Best regard ~ Donald
